Oracle feature (#75)
autonity/docs.autonity.org: content/docs/concepts/oracle-network/_index.md

Reviewer: In this file ("Oracle network"), the following three sections can be moved to `content/docs/concepts/oracle-server/_index.md` instead:

- Data adaptors - plugin architecture
- Oracle data providers
- Oracle server lifecycle

Author: Moved in commit [4f47b0f](https://github.com/autonity/docs.autonity.org/pull/75/commits/4f47b0f014f27d9c3c8fd868501c7f75a48d94e5).

I was a little sceptical, but this reorganisation definitely reads better, great!

Prerequisites for participating in the oracle network as a validator node operator are:

- An [Autonity Go Client](/concepts/client/) full node installation, joined to an Autonity network.
- An [Autonity Oracle Server](/concepts/oracle-server/) installation, configured to connect to your full node and external price data sources via plugin adaptors.
- Plugin adaptors for connecting to external data provider sources to retrieve price data for the currency pairs required for your node's connected Autonity network.

## Oracle identity, accounts and keypairs
The oracle server makes use of a single account and private/public [key pair](/glossary/#key-pair) for oracle identity, oracle protocol participation, and cryptographic security. 

### Oracle server key

The private/public key pair of the oracle server node.

The private key is used:

- By an oracle server for submitting price report transactions to the Oracle Contract on-chain.
- To generate the `proof` of enode ownership required for validator registration. The `proof` is generated using the [`genOwnershipProof`](/reference/cli/agc/#command-line-options) command-line option of the Autonity Go Client. 

> 

Reviewer 2: It confuse me here, in my understanding, the enode proof are generated by using the autonity client's node key which is different comparing to the private key being used by oracle server. They are different keys.

Author: Ah, it uses both node key and oracke ley now `/validators/register-vali/#step-1-generate-a-cryptographic-proof-of-node-ownership`:

```
autonity genOwnershipProof --nodekey <NODE_KEY_PATH> --oraclekey <ORACLE_KEY_PATH> <TREASURY_ACCOUNT_ADDRESS>
```

I've edited this to clarify

Author: Can this be resolved @Jason-Zhangxin-Chen ?

Author: @Jason-Zhangxin-Chen

Reviewer 2: > @Jason-Zhangxin-Chen

LGTM.

The protocol computes median price by aggregation:

- Off-chain _L1 aggregation_. Oracle Server's collect source price data for currency pairs from one or more sources. If data is taken from multiple sources, then an aggregation is performed to consolidate price data into a standardised report. This price report (or price data 'sample') is then pushed to the oracle contract deployed on the Autonity L1 network via a transaction submitted from the oracle server account.
- On-chain _L2 aggregation_. The oracle contract aggregates the supplied price data reports to compute a reference exchange rate data through a voting mechanism using a [commit-reveal](/concepts/oracle-network/#commit-and-reveal) protocol in [voting rounds](/concepts/oracle-network/#voting-rounds). 

> The commit and reveal scheme is applied for security in price voting and  results in a round lag between oracle price updates.

Reviewer 2: duplicated spaces before the world `results in`

Author: Well spotted! Fixed in commit [3b4b466](https://github.com/autonity/docs.autonity.org/pull/75/commits/3b4b466b587da4cd3507a0ffbf068f0b9191b6e8)

---
title: "Oracle network"
linkTitle: "Oracle network"
weight: 4
description: >
  Autonity oracle network: the role of oracles, the functions they perform, and their lifecycle
---

## Overview

This section describes the oracle protocol, the role of validators in providing and agreeing median price data, and the lifecycle for oracle voting and currency pair management.

> Autonity provides consensus-computed median price data for selected currency pairs as an L1 platform feature. The submission of price data for aggregation and voting is a validator responsibility and executed according to an _oracle protocol_. As consequence, validator nodes form an _oracle network_. When a member of the [consensus committee](/glossary/#consensus-committee), validators must submit raw price data reports for currency pairs on-chain, and participate in the oracle protocol to agree an aggregated median price for the submitted price data in discrete _voting rounds_ using a _commit and reveal_ protocol. As voting rounds complete, the aggregated median price data computed and provided by the Autonity network is updated. As [epoch](/glossary/#epoch)'s end the set of oracle voters changes to accord with new consensus committee membership.

Reviewer: ```suggestion
Autonity provides consensus-computed median price data for selected currency pairs as an L1 platform feature. The submission of price data for aggregation and voting is a validator responsibility and executed according to an _oracle protocol_. As consequence, validator nodes form an _oracle network_. When serving as a member of the [consensus committee](/glossary/#consensus-committee), validators must submit raw price data reports for currency pairs on-chain, and participate in the oracle protocol to agree an aggregated median price for the submitted price data in discrete _voting rounds_ using a _commit and reveal_ protocol. As voting rounds complete, the aggregated median price data computed and provided by the Autonity network is updated. As [epoch](/glossary/#epoch)'s end the set of oracle voters changes to accord with new consensus committee membership.
```

- As the `msg.Sender` address used by the oracle server to submit transactions for oracle price report submissions by function call to the Oracle Contract on-chain.

### Oracle identifier

A unique identifier for the Autonity Oracle Server used as the oracle identity in validator registration, where it is provided as an  `oracleAddress` parameter.

The identity is created as an ethereum format account address, generated by the entity operating the oracle server as part of the deployment process. It provides an unambiguous relationship between oracle server identity and the associated validator node for which the oracle server is a data provider.

{{< alert title="Note" >}}
The oracle identifier is stored in the `oracleAddress` field in the Validator data struct maintained in state.

Note that this identifier is the _oracle server's_ on-chain identity and is distinct from the `nodeAddress` which is the validator node identifier, and the `treasury` account which is the validator _operator's_ account. 
The identifier may be referred to as the `oracle` account.
{{< /alert >}}

## Oracle network

> Validator nodes must participate in the [oracle protocol](/concepts/oracle-network/#oracle-protocol) when a member of the consensus committee.

Reviewer: ```suggestion
Validator nodes must participate in the [oracle protocol](/concepts/oracle-network/#oracle-protocol) when serving as a member of the consensus committee.
```

Reviewer: @cmjc

## Oracle protocol

The computation of median price data is governed by protocol logic and rules. The protocol has off- and on-chain operations. Price data is collected from external data providers by [oracle servers](/glossary/#autonity-oracle-server-aos) run by validator operators and submitted on-chain to an oracle [contract](/glossary/#smart-contract) that computes aggregate median price data in [voting rounds](/glossary/#voting-round). Validators are responsible for providing source data from off-chain and voting to agree a median price on-chain when members of the consensus committee.

The protocol computes median price by aggregation:

- Off-chain _L1 aggregation_. Oracle Server's collect source price data for currency pairs from one or more sources. If data is taken from multiple sources, then an aggregation is performed to consolidate price data into a standardised report. This price report (or price data 'sample') is then pushed to the oracle contract deployed on the Autonity L1 network via a transaction submitted from the oracle server account.

> - On-chain _L2 aggregation_. The oracle contract aggregates the supplied price data reports to compute a reference exchange rate data through a voting mechanism using a [commit-reveal](/concepts/oracle-network/#commit-and-reveal) protocol in [voting rounds](/concepts/oracle-network/#voting-rounds). 

Reviewer: Let's not use the term L2 to refer to EVM transactions, as L2 has a different meaning in the space. For example: <https://chain.link/education-hub/what-is-layer-2>

Reviewer: @cmjc

Author: @aiman - Terminology revised in commit 3bb31cb6e2f93c7b4ac3b5bc545f4096802f6a0f

### Voter selection

Oracle voters are selected from the current consensus committee and are kept in sync with consensus committee changes by protocol as the committee changes with new epochs.

Voters are allowed to vote once each round. If a validator leaves the  consensus committee then their vote is discarded; if a validator joins the consensus committee then their first vote is not taken into account.

### Commit and reveal

A _commit and reveal_ protocol is used for price voting. In each round the oracle voter submits a vote transaction providing:

- `_commit`: a cryptographic hash of the new price report submitted for the round
- `_reports`: a reveal of the price report submitted for the previous round
- `_salt`: a salt value used to generate the previous round's commitment

> Protocol then uses the _salt_ to verify the previous round's _commit_ matches the _reveal_ for that commit provided by the voter in the current round. This introduces a time lag in oracle price updates: prices are submitted in round `1`, revealed and median price data at the end of round `2`.

Reviewer: ```suggestion
Protocol then uses the _salt_ to verify the previous round's _commit_ matches the _reveal_ for that commit provided by the voter in the current round. This introduces a time lag in oracle price updates: prices are submitted in round `1`, revealed during round `2`, with median price data computed at the end of round `2`.
```

Reviewer: @cmjc

To exemplify oracle frequency of new median price data publication from genesis in a scenario of vote period set to 30 blocks:

|Time (Block Height) | Event|
|:----:|:-----|
|`Genesis` | Network genesis; initiation of first oracle voting round: `R1` |

> |`Block 1...29`| Oracle voters submit price report vote, providing _commit_; _reveal_ and _salt_ null. |

Reviewer: Not so sure about `29` being the cut-off. As far as I understand, a price submission could be included in block `30`, before the `finalize()` executes. We may need to double-check the block numbers used in this table.

Also, I think the focus in descriptions should focus more on the inclusion of the oracle's vote in a block, and not on their submission time. This is because even if I submit a vote at block `B`, it may not be included until block `B+t`, where `t` can be any number of blocks. So the wording should be on the inclusion in block, not on submission.

Reviewer: @cmjc

Author: You are correct, it should be `30` and *not* `29`.

Author: Fixed in commit cbd2017.

|`Block 30`| `R1` voting round ends. No previous round so no _commits_ to _reveal_ and no valid price submissions for median price computation for `R1`. New Round event emitted for `R2`. |
|`Block 31`| New Round begins: `R2` |
|`Block 31...59`| Oracle voters submit price report vote, providing _commit_ for `R2`; _reveal_ and _salt_ for `R1`. |
|`Block 60`| `R2` voting round ends. Commit and reveal; median price round data computed for `R2`. New Round event emitted for `R3`&hellip; |

{{% alert title="Note" %}}
If an oracle fails to vote in a round, or the reveal does not match the past commit, then the oracle voter is recorded as submitting an invalid price for the round.
{{% /alert %}}

### Currency pair management

> The currency pair symbols for which the oracle network provides price data is set at network genesis in the [genesis configuration file](/reference/genesis/#genesis-configuration-file) in the `symbols` field; see `[config.autonity.oracle](/reference/genesis/#configautonityoracle-object)` object.

Reviewer: The `config.autonity.oracle` URL is broken here.

Reviewer: @cmjc still broken

Author: Caused by back tick markers. Fixed in commit 145273a

The currency pair symbols set for a network can be returned by a contract call using the [getSymbols()](/reference/api/oracle/#getsymbols) function.

> The pairs can be updated post-genesis by a governance-only function, `[setSymbols()](/reference/api/aut/op-prot/#setsymbols-oracle-contract).

Reviewer: ```suggestion
The pairs can be updated post-genesis by a governance-only function, [setSymbols()](/reference/api/aut/op-prot/#setsymbols-oracle-contract).
```

Reviewer: @cmjc or some other formatting

Author: Thanks! I've fixed in commit 145273a

- Simulator plugin: for simulated ATN/NTN data. See the `simulator_plugin` adaptor in [`/plugins`<i class='fas fa-external-link-alt'></i>](https://github.com/autonity/autonity-oracle/tree/master/plugins). 

{{% alert title="Info" color="info"%}}
ATN and NTN symbols are preview listed but untraded:

- https://www.coingecko.com/en/coins/auton
- https://www.coingecko.com/en/coins/newton

Plugins for retrieving ATN/NTN price data are to be developed for Mainnet launch.
{{% /alert %}}

### Developing data plugins
Additional data adaptors for any external data source can be developed using the oracle server's plugin template. See:

- Adaptor code template `template_plugin` in [`/plugins`<i class='fas fa-external-link-alt'></i>](https://github.com/autonity/autonity-oracle/tree/master/plugins).
- Guide for how _To write a new plugin_ using the template in [`/plugins/README`<i class='fas fa-external-link-alt'></i>](https://github.com/clearmatics/autonity-oracle/tree/master/plugins#readme).

## Oracle data consumers

Primary consumers of oracle data are:

> - Autonity Stabilisation Mechanism

Reviewer: ```suggestion
- Auton Stabilisation Mechanism
```

Reviewer: @cmjc

> - DApps deployed on the Autonity L1 network can access median price data via the oracle contract interface.

Reviewer: I would just call them smart contracts:

```suggestion
- Smart contracts deployed on the Autonity L1 network can access median price data via the oracle contract interface.
```

Reviewer: @cmjc
